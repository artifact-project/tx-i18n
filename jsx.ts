import { i18nJSXFactory } from './src/jsx/jsx';
import { createElement } from 'react';

export default i18nJSXFactory(createElement);
