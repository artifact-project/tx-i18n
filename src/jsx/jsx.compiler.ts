import { Plural } from '@artifact-project/i18n';
import { icu, Token } from '../utils/icu';
import { AST, Element } from 'format-message-parse';
import { htmlDecode } from '../utils/entities';
import { InterpolateString, createInterpolateString, initInterpolateString } from '../i18n/i18n.compiler';

const EMPTY_ARRAY = [];
const EMPTY_STRING = '';
const HASH_BANG = '#';

const COMPILED = {} as Record<string, (parts: any[]) => LikeJSXElement>;

export type LikeJSXElement = {
	type: string | Function;
	props: {
		children: Array<LikeJSXElement | string>;
		[key:string]: any;
	};
}

export type LikeJSXFactory = (type: string | Function, props: object, ...children: any[]) => LikeJSXElement;

export function jsxCompile(
	phrase: string,
	createElement: LikeJSXFactory,
	plural?: Plural<any, any>,
): (parts: any[]) => LikeJSXElement {
	if (COMPILED.hasOwnProperty(phrase)) {
		return COMPILED[phrase];
	}

	const icuTools = icu.fn({ plural });
	const tokenCache = new Map<Element, {
		type: Token;
		idx: number | null;
		raw?: string;
		plural: InterpolateString | null;
	}>()
	

	let tokens = [] as AST;
	let compiled = null;

	try {
		tokens = icu.parse(phrase, {
			tags: true,
		});
	} catch (err) {
		tokens = [phrase];
		console.error(err);
	}

	// Tokens processing
	// console.log(tokens);
	function processing(values: any, tokens: AST, idx: number) {
		const el = values[idx];
		const args = [el.type, el.props];
		// write(`($$$ = __vls__[${idx}], __CREATE__($$$.type, $$$.props`);

		for (let i = 0; i < tokens.length; i++) {
			const token = tokens[i];
			const cache = tokenCache.get(token) || {
				type: icu.getTokenType(token),
				idx: null,
				plural: null,
			};

			switch (cache.type) {
				case Token.String:
					if (token !== EMPTY_STRING) {
						const raw = cache.raw || htmlDecode(token as string);
						cache.raw = raw;
						args.push(raw);
					}
					break;

				case Token.Simple:
					idx = cache.idx === null ? icu.parseTokenIdx(token, 0) : cache.idx;
					cache.idx = idx;
					args.push(token[0] === HASH_BANG ? values[idx]._ : values[idx]);
					break;

				case Token.Tag:
					idx = cache.idx === null ? icu.parseTokenIdx(token, 0) : cache.idx;
					cache.idx = idx;

					if (token.length === 3) {
						args.push(processing(values, (token[2] as any).children, idx));
					} else {
						args.push(processing(values, EMPTY_ARRAY, idx));
					}
					break;

				case Token.Typed:
					throw new Error("todo: Token.Typed");

				case Token.Styled:
					throw new Error("todo: Token.Typed");

				case Token.Plural:
					if (cache.plural === null) {
						cache.plural = createInterpolateString(icuTools);
						initInterpolateString(cache.plural, [token], 0);
					}

					cache.plural.value = null;
					cache.plural.values = values;

					args.push(cache.plural.toString());
					break;
			}

			tokenCache.set(token, cache);
		}

		return args.length === 2
			? createElement(el.type, el.props)
			: args.length === 2
			? createElement(el.type, el.props. args[2])
			: createElement.apply(null, args)
	}

	// Compile
	// console.log(code);
	try {
		compiled = function (values: any) {
			return processing(values, tokens, 0)
		}
	} catch (err) {
		compiled = () => phrase;
		console.warn(`Parse failed: "${phrase}"`);
		console.error(err);
	}

	COMPILED[phrase] = compiled;
	return compiled;
}
