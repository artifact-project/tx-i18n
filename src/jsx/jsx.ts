import { getTranslate } from '../i18n/locale';

type JsxElement = {
	type: string;
	props: {
		children: Array<JsxElement | string>;
		[key:string]: any;
	};
}

const R_SEP = /<(\/|#)?(\d+)(\/?)>/g;
const R_CLEANER = /,(\n\s*\))/g;
const COMPILED = {};

function compile(input: string) {
	const source = [];
	const tokens = input.split(R_SEP);
	let indent = '';

	function line(code: string, ind?: number) {
		if (ind < 0) {
			indent = indent.slice(2);
		}

		source.push(`${indent}${code}`);

		if (ind > 0) {
			indent += '  ';
		}
	}

	line(`__CREATE__(`, +1);
	line('__parts__[0].type,');
	line('__parts__[0].props,');

	for (let i = 0; i < tokens.length; i += 4) {
		const text = tokens[i];
		const type = tokens[i + 1];
		const partIdx = tokens[i + 2];
		const selfClosed = tokens[i + 3];

		if (text !== '') {
			line(`${JSON.stringify(text)},`);
		}

		if (type === '#') {
			line(`__parts__[${partIdx}],`);
		} else if (selfClosed) {
			line(`__CREATE__(__parts__[${partIdx}].type, __parts__[${partIdx}].props),`);
		} else if (type === '/') {
			line('),', -1);
		} else if (partIdx) {
			line(`__CREATE__(`, +1);
			line(`__parts__[${partIdx}].type,`);
			line(`__parts__[${partIdx}].props,`);
		}
	}

	line(')', -1);

	return Function(`__CREATE__, __parts__`, `return (${source.join('\n').replace(R_CLEANER, '$1')});`);
}

export function jsxFactory(create: (type: string, props: object, ...children: any[]) => JsxElement) {
	return function jsx(phrase: string, parts) {
		const translatePhrase = getTranslate(phrase);

		if (!COMPILED.hasOwnProperty(translatePhrase)) {
			COMPILED[translatePhrase] = compile(translatePhrase);
		}

		return COMPILED[translatePhrase](create, parts);
	};
}