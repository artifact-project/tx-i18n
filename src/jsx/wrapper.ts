import * as React from 'react';
import { addLangObserver } from '../i18n/locale';

export class Wrapper extends React.Component<{value: Function}> {
	private unsubscribe: () => void;

	componentDidMount() {
		this.unsubscribe = addLangObserver(() => {
			this.forceUpdate();
		});
	}

	componentWillUnmount() {
		this.unsubscribe();
	}

	shouldComponentUpdate() {
		return false;
	}

	render() {
		return this.props.value();
	}
}
