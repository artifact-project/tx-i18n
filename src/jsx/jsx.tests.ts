import { jsxFactory } from './jsx';

it('jsx', () => {
	const jsx = jsxFactory((type, props, ...children) => ({
		type,
		props: {
			...Object(props),
			children,
		},
	}));

	expect(jsx(
		"Привет <#1>, нажми <2>здесь</2> чтобы продолжить<3/>.",
		[
			{type: 'div', props: null},
			"Вася",
			{type: "a", props: {href: "#"}},
			{type: "i", props: {className: "icon"}},
		]
	)).toMatchSnapshot();
});
