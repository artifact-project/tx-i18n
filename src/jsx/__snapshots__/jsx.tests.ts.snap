// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx 1`] = `
Object {
  "props": Object {
    "children": Array [
      "Привет ",
      "Вася",
      ", нажми ",
      Object {
        "props": Object {
          "children": Array [
            "здесь",
          ],
          "href": "#",
        },
        "type": "a",
      },
      " чтобы продолжить",
      Object {
        "props": Object {
          "children": Array [],
          "className": "icon",
        },
        "type": "i",
      },
      ".",
    ],
  },
  "type": "div",
}
`;
